04-evm-mapping.md

```diff
@@ -0,0 +1,307 @@
+# FVM <> EVM mapping
+
+This document describes the mappings of structures, types, and procedures between the native FVM environment (WASM), and the EVM foreign VM (emulated).
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*
+
+- [Integration](#integration)
+  - [State tree](#state-tree)
+  - [Mechanics and interfaces](#mechanics-and-interfaces)
+- [Contract memory model](#contract-memory-model)
+- [Account storage model](#account-storage-model)
+- [Addressing scheme](#addressing-scheme)
+  - [Proposed solution: universal stable addresses](#proposed-solution-universal-stable-addresses)
+- [Gas accounting and execution halt semantics](#gas-accounting-and-execution-halt-semantics)
+- [Ethereum logs/events](#ethereum-logsevents)
+- [Blockchain timing](#blockchain-timing)
+- [Precompiles](#precompiles)
+- [Cryptographic primitives](#cryptographic-primitives)
+- [Chain-specific opcodes](#chain-specific-opcodes)
+- [Cross-contract calls](#cross-contract-calls)
+- [References](#references)
+- [Annex A: Addressing solutions considered](#annex-a-addressing-solutions-considered)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Integration
+
+### State tree
+
+There won't be a segregated world state for Ethereum actors in the Filecoin blockchain. We will not use the Patricia Merkle Trie structure. Instead, EVM account state will be folded and docked into the Filecoin state tree.
+
+### Mechanics and interfaces
+
+EVM smart contracts (also known as EVM foreign actors in Filecoin terminology), are represented this way in the Filecoin state tree:
+
+```
+{
+	Code:       cid('fil/<version>/evm/<runtime_version>'),
+	Head:       cid(<actor_state>),
+	Nonce:      uint64(<nonce>),   
+	Balance:    bigint(<fil_balance>),
+}
+```
+
+Notice that EVM foreign actors are typed in the state tree with a CodeCID that does not correspond to their EVM bytecode. Instead, the CodeCID points to the WASM bytecode of the **EVM foreign runtime**.
```

Comment:
This should just be the CID of the EVM itself. We currently use strings, but we only do that because the code doesn't live on-chain.

Reply:
Yep, that's what I meant to say but it was a long-winded way of doing so. I'll adjust the text.